src/coreComponents/constitutive/relativePermeability/TableRelativePermeabilityHysteresis.cpp

```diff
@@ -0,0 +1,706 @@
+    {
+      TableFunction const & drainageRelPermTable = functionManager.getGroup< TableFunction >( m_drainageNonWettingIntermediateRelPermTableNames[ip] );
+      m_drainageRelPermKernelWrappers.emplace_back( drainageRelPermTable.createKernelWrapper() );
+    }
+
+    TableFunction const & imbibitionWettingRelPermTable = m_phaseHasHysteresis[IPT::WETTING]
+      ? functionManager.getGroup< TableFunction >( m_imbibitionWettingRelPermTableName )
+      : functionManager.getGroup< TableFunction >( m_drainageWettingIntermediateRelPermTableNames[0] );
+    m_imbibitionRelPermKernelWrappers.emplace_back( imbibitionWettingRelPermTable.createKernelWrapper() );
+
+    TableFunction const & imbibitionNonWettingRelPermTable = m_phaseHasHysteresis[IPT::NONWETTING]
+      ? functionManager.getGroup< TableFunction >( m_imbibitionNonWettingRelPermTableName )
+      : functionManager.getGroup< TableFunction >( m_drainageNonWettingIntermediateRelPermTableNames[0] );
+    m_imbibitionRelPermKernelWrappers.emplace_back( imbibitionNonWettingRelPermTable.createKernelWrapper() );
+  }
+
+}
+
+TableRelativePermeabilityHysteresis::KernelWrapper
+TableRelativePermeabilityHysteresis::createKernelWrapper()
+{
+
+  // we want to make sure that the wrappers are always up-to-date, so we recreate them everytime
+  createAllTableKernelWrappers();
+
+  // then we create the actual TableRelativePermeabilityHysteresis::KernelWrapper
+  return KernelWrapper( m_drainageRelPermKernelWrappers,
+                        m_imbibitionRelPermKernelWrappers,
+                        m_jerauldParam_a,
+                        m_jerauldParam_b,
+                        m_killoughCurvatureParam,
+                        m_phaseHasHysteresis,
+                        m_landParam,
+                        m_drainagePhaseMinVolFraction,
+                        m_imbibitionPhaseMinVolFraction,
+                        m_drainagePhaseMaxVolFraction,
+                        m_imbibitionPhaseMaxVolFraction,
+                        m_drainagePhaseRelPermEndPoint,
+                        m_imbibitionPhaseRelPermEndPoint,
+                        m_phaseTypes,
+                        m_phaseOrder,
+                        m_phaseMinHistoricalVolFraction,
+                        m_phaseMaxHistoricalVolFraction,
+                        m_phaseRelPerm,
+                        m_dPhaseRelPerm_dPhaseVolFrac );
+}
+
+void TableRelativePermeabilityHysteresis::resizeFields( localIndex const size, localIndex const numPts )
+{
+  RelativePermeabilityBase::resizeFields( size, numPts );
+
+  integer const numPhases = numFluidPhases();
+
+  m_phaseMaxHistoricalVolFraction.resize( size, numPhases );
+  m_phaseMinHistoricalVolFraction.resize( size, numPhases );
+}
+
+void TableRelativePermeabilityHysteresis::initializePhaseVolFractionState( arrayView2d< real64 const, compflow::USD_PHASE > const & initialPhaseVolFraction ) const
```

Comment:
It seems to me `initializePhaseVolFractionState` and `saveConvergedPhaseVolFractionState` are doing the same thing, essentially. Assuming `m_phaseMinHistoricalVolFraction` is initialized to `1.0` and `m_phaseMaxHistoricalVolFraction` is initialized to `0.0`, both computations amount to
```c++
minHistS[ip] = min( minHistS[ip], currentS[ip] );
maxHistS[ip] = max( maxHistS[ip], currentS[ip] );
```
I would just keep `saveConvergedPhaseVolFractionState` and call it when solver is initializing the models with initial saturation.

Reply:
On a second thought, maybe this does not generalize well to all types of models. I'm struggling to come up with an example, but maybe are models for which "first time seeing state" and "update previously seen state" is a meaningful distinction that the model cannot figure out on its own, which necessitates having two functions in the interface.